// @translate-tools/core/utils/Semaphore.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Semaphore = void 0;
var __awaiter = void 0 && (void 0).__awaiter || function (thisArg, _arguments, P, generator) {
  function adopt(value) {
    return value instanceof P ? value : new P(function (resolve) {
      resolve(value);
    });
  }
  return new (P || (P = Promise))(function (resolve, reject) {
    function fulfilled(value) {
      try {
        step(generator.next(value));
      } catch (e) {
        reject(e);
      }
    }
    function rejected(value) {
      try {
        step(generator["throw"](value));
      } catch (e) {
        reject(e);
      }
    }
    function step(result) {
      result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected);
    }
    step((generator = generator.apply(thisArg, _arguments || [])).next());
  });
};
var __generator = void 0 && (void 0).__generator || function (thisArg, body) {
  var _ = {
      label: 0,
      sent: function () {
        if (t[0] & 1) throw t[1];
        return t[1];
      },
      trys: [],
      ops: []
    },
    f,
    y,
    t,
    g;
  return g = {
    next: verb(0),
    "throw": verb(1),
    "return": verb(2)
  }, typeof Symbol === "function" && (g[Symbol.iterator] = function () {
    return this;
  }), g;
  function verb(n) {
    return function (v) {
      return step([n, v]);
    };
  }
  function step(op) {
    if (f) throw new TypeError("Generator is already executing.");
    while (g && (g = 0, op[0] && (_ = 0)), _) try {
      if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
      if (y = 0, t) op = [op[0] & 2, t.value];
      switch (op[0]) {
        case 0:
        case 1:
          t = op;
          break;
        case 4:
          _.label++;
          return {
            value: op[1],
            done: false
          };
        case 5:
          _.label++;
          y = op[1];
          op = [0];
          continue;
        case 7:
          op = _.ops.pop();
          _.trys.pop();
          continue;
        default:
          if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) {
            _ = 0;
            continue;
          }
          if (op[0] === 3 && (!t || op[1] > t[0] && op[1] < t[3])) {
            _.label = op[1];
            break;
          }
          if (op[0] === 6 && _.label < t[1]) {
            _.label = t[1];
            t = op;
            break;
          }
          if (t && _.label < t[2]) {
            _.label = t[2];
            _.ops.push(op);
            break;
          }
          if (t[2]) _.ops.pop();
          _.trys.pop();
          continue;
      }
      op = body.call(thisArg, _);
    } catch (e) {
      op = [6, e];
      y = 0;
    } finally {
      f = t = 0;
    }
    if (op[0] & 5) throw op[1];
    return {
      value: op[0] ? op[1] : void 0,
      done: true
    };
  }
};
/**
 * Semaphore for the flow control in queues
 *
 * @example
 * const semafor = QueueSemafor({ timeout: 100 });
 * items.map(async item=> {
 * 	const free = await semafor.take();
 * 	// do something with item...
 * 	free();
 * })
 */
var Semaphore = /** @class */function () {
  function Semaphore(options) {
    this.timeout = 0;
    this.hijackPrevention = true;
    this.wait = function (time) {
      return new Promise(function (res) {
        return setTimeout(res, time);
      });
    };
    this.lastAccess = 0;
    this.semafor = null;
    var _a = options || {},
      timeout = _a.timeout,
      hijackPrevention = _a.hijackPrevention;
    if (timeout !== undefined) {
      if (timeout < 0) {
        throw new Error('Negative number');
      }
      this.timeout = timeout;
    }
    if (hijackPrevention !== undefined) {
      this.hijackPrevention = hijackPrevention;
    }
  }
  Semaphore.prototype.take = function () {
    return __awaiter(this, void 0, void 0, function () {
      var idle, semaforResolver;
      var _this = this;
      return __generator(this, function (_a) {
        switch (_a.label) {
          case 0:
            if (!true) return [3 /*break*/, 6];
            if (!(this.timeout > 0)) return [3 /*break*/, 2];
            idle = new Date().getTime() - this.lastAccess;
            if (!(idle < this.timeout)) return [3 /*break*/, 2];
            return [4 /*yield*/, this.wait(this.timeout - idle)];
          case 1:
            _a.sent();
            _a.label = 2;
          case 2:
            if (this.semafor === null) {
              return [3 /*break*/, 6];
            }
            return [4 /*yield*/, this.semafor];
          case 3:
            _a.sent();
            if (!this.hijackPrevention) return [3 /*break*/, 5];
            return [4 /*yield*/, this.wait(Math.floor(Math.random() * 30))];
          case 4:
            _a.sent();
            _a.label = 5;
          case 5:
            return [3 /*break*/, 0];
          case 6:
            this.semafor = new Promise(function (resolve) {
              semaforResolver = resolve;
            });
            return [2 /*return*/, function () {
              _this.lastAccess = new Date().getTime();
              _this.semafor = null;
              semaforResolver();
            }];
        }
      });
    });
  };
  return Semaphore;
}();
exports.Semaphore = Semaphore;
